src/osmdata-data_frame.cpp

```diff
@@ -0,0 +1,259 @@
+    } else
+        kv_df = R_NilValue;
+
+    return kv_df;
+}
+
+//' get_osm_ways
+//'
+//' Store key-val pairs for OSM ways as a list/data.frame
+//'
+//' @param kv_df Pointer to Rcpp::DataFrame to hold key-value pairs
+//' @param way_ids Vector of <osmid_t> IDs of ways to trace
+//' @param ways Pointer to all ways in data set
+//' @param nodes Pointer to all nodes in data set
+//' @param unique_vals pointer to all unique values (OSM IDs and keys) in data set
+//' 
+//' @noRd 
+void osm_df::get_osm_ways (Rcpp::DataFrame &kv_df,
+        const std::set <osmid_t> &way_ids, const Ways &ways,
+        const UniqueVals &unique_vals)
+{
+
+    size_t nrow = way_ids.size (), ncol = unique_vals.k_way.size ();
+    std::vector <std::string> waynames;
+    waynames.reserve (way_ids.size ());
+
+    Rcpp::CharacterMatrix kv_mat (Rcpp::Dimension (nrow, ncol));
+    std::fill (kv_mat.begin (), kv_mat.end (), NA_STRING);
+    unsigned int count = 0;
+    for (auto wi = way_ids.begin (); wi != way_ids.end (); ++wi)
+    {
+        Rcpp::checkUserInterrupt ();
+        waynames.push_back (std::to_string (*wi));
+        auto wj = ways.find (*wi);
+        osm_convert::get_value_mat_way (wj, unique_vals, kv_mat, count);
+        count++;
+    }
+
+    kv_df = R_NilValue;
+    if (way_ids.size () > 0)
+    {
+        kv_mat.attr ("names") = unique_vals.k_way;
+        kv_mat.attr ("dimnames") = Rcpp::List::create (waynames, unique_vals.k_way);
+        if (kv_mat.nrow () > 0 && kv_mat.ncol () > 0)
+            kv_df = osm_convert::restructure_kv_mat (kv_mat, false);
+    }
+}
+
+//' get_osm_nodes
+//'
+//' Store OSM nodes as `sf::POINT` objects
+//'
+//' @param kv_df Pointer to Rcpp::DataFrame to hold key-value pairs
+//' @param nodes Pointer to all nodes in data set
+//' @param unique_vals pointer to all unique values (OSM IDs and keys) in data set
+//' 
+//' @noRd 
+void osm_df::get_osm_nodes (Rcpp::DataFrame &kv_df,
+        const Nodes &nodes, const UniqueVals &unique_vals)
+{
+    size_t nrow = nodes.size (), ncol = unique_vals.k_point.size ();
+
+    Rcpp::CharacterMatrix kv_mat (Rcpp::Dimension (nrow, ncol));
+    std::fill (kv_mat.begin (), kv_mat.end (), NA_STRING);
+
+    std::vector <std::string> ptnames;
+    ptnames.reserve (nodes.size ());
+    unsigned int count = 0;
+    for (auto ni = nodes.begin (); ni != nodes.end (); ++ni)
+    {
+        if (count % 1000 == 0)
+            Rcpp::checkUserInterrupt ();
+
+        ptnames.push_back (std::to_string (ni->first));
+        for (auto kv_iter = ni->second.key_val.begin ();
+                kv_iter != ni->second.key_val.end (); ++kv_iter)
+        {
+            const std::string &key = kv_iter->first;
+            unsigned int ndi = unique_vals.k_point_index.at (key);
+            kv_mat (count, ndi) = kv_iter->second;
+        }
+        count++;
+    }
+    if (unique_vals.k_point.size () > 0)
+    {
+        kv_mat.attr ("dimnames") = Rcpp::List::create (ptnames, unique_vals.k_point);
```

Comment:
```suggestion
        kv_mat.attr ("names") = unique_vals.k_point;
```
row names are not relevant a can be removed